tests/integration_tests/tests/timeout.rs

```diff
@@ -0,0 +1,94 @@
+use integration_tests::pb::{test_client, test_server, Input, Output};
+use std::time::Duration;
+use tokio::net::TcpListener;
+use tonic::{transport::Server, Code, Request, Response, Status};
+
+#[tokio::test]
+#[ignore]
+async fn cancelation_on_timeout() {
+    struct Svc;
+
+    #[tonic::async_trait]
+    impl test_server::Test for Svc {
+        async fn unary_call(&self, _req: Request<Input>) -> Result<Response<Output>, Status> {
+            // Wait for a time longer than the timeout
+            tokio::time::sleep(Duration::from_millis(1_000)).await;
+            Ok(Response::new(Output {}))
+        }
+    }
+
+    let svc = test_server::TestServer::new(Svc);
+
+    let listener = TcpListener::bind("127.0.0.1:0").await.unwrap();
+    let addr = listener.local_addr().unwrap();
+
+    tokio::spawn(async move {
+        Server::builder()
+            .add_service(svc)
+            .serve_with_incoming(tokio_stream::wrappers::TcpListenerStream::new(listener))
+            .await
+            .unwrap();
+    });
+
+    let mut client = test_client::TestClient::connect(format!("http://{}", addr))
+        .await
+        .unwrap();
+
+    let mut req = Request::new(Input {});
+    req.metadata_mut()
+        .insert("grpc-timeout", "500m".parse().unwrap());
+
+    let res = client.unary_call(req).await;
+    dbg!(&res);
+
+    let err = res.unwrap_err();
+    assert!(err.message().contains("Timeout expired"));
+
+    // TODO(david): make this work. Will require mapping `TimeoutExpired` errors into
+    // `Code::Cancelled` but can't quite figure out how to do that.
+    assert_eq!(err.code(), Code::Cancelled);
```

Comment:
@LucioFranco do you know how to fix this? Basically our new `TimeoutExpired` error should change the status code to `Code::Cancelled`. It gets mapped to `Unknown` currently. Did some digging but couldn't quite figure out where that happens.

Reply:
Still not sure how to handle this. @LucioFranco do you have idea?

Reply:
I pulled the code down locally its here

```
diff --git a/tonic/src/status.rs b/tonic/src/status.rs
index dcf381d..309fe0c 100644
--- a/tonic/src/status.rs
+++ b/tonic/src/status.rs
@@ -331,6 +331,10 @@ impl Status {
                 if let Some(h2) = err.downcast_ref::<h2::Error>() {
                     return Some(Status::from_h2_error(h2));
                 }
+
+                if let Some(timeout) = err.downcast_ref::<crate::transport::TimeoutExpired>() {
+                    return Some(Status::cancelled(format!("{}", timeout)));
+                }
             }

             cause = err.source();
```

This was the fix for me.